File: tests/integration/games/checkers.py
import random

# Initialize the board
def initialize_board():
    board = [[' ' for _ in range(8)] for _ in range(8)]
    for row in range(3):
        for col in range(8):
            if (row + col) % 2 == 1:
                board[row][col] = 'C'  # Computer pieces
    for row in range(5, 8):
        for col in range(8):
            if (row + col) % 2 == 1:
                board[row][col] = 'U'  # User pieces
    return board

# Print the board
def print_board(board):
    print("  " + " ".join(str(i) for i in range(8)))
    for i, row in enumerate(board):
        print(f"{i} " + " ".join(row))

# Check if a move is valid
def is_valid_move(board, player, start, end):
    sx, sy = start
    ex, ey = end
    if ex < 0 or ex >= 8 or ey < 0 or ey >= 8:
        return False
    if board[ex][ey] != ' ':
        return False
    if player == 'U' and ex == sx - 1 and abs(ey - sy) == 1:
        return True
    if player == 'C' and ex == sx + 1 and abs(ey - sy) == 1:
        return True
    return False

# Make a move
def make_move(board, start, end):
    sx, sy = start
    ex, ey = end
    board[ex][ey] = board[sx][sy]
    board[sx][sy] = ' '

# Get all valid moves for a player
def get_valid_moves(board, player):
    moves = []
    for x in range(8):
        for y in range(8):
            if board[x][y] == player:
                for dx, dy in [(-1, -1), (-1, 1), (1, -1), (1, 1)]:
                    if is_valid_move(board, player, (x, y), (x + dx, y + dy)):
                        moves.append(((x, y), (x + dx, y + dy)))
    return moves

# Computer's turn
def computer_turn(board):
    moves = get_valid_moves(board, 'C')
    if not moves:
        return False
    move = random.choice(moves)
    make_move(board, move[0], move[1])
    return True

# User's turn
def user_turn(board):
    moves = get_valid_moves(board, 'U')
    if not moves:
        return False
    print("Your turn! Enter your move as 'start_x start_y end_x end_y'")
    while True:
        try:
            move = input("Enter your move: ").strip().split()
            start = (int(move[0]), int(move[1]))
            end = (int(move[2]), int(move[3]))
            if is_valid_move(board, 'U', start, end):
                make_move(board, start, end)
                return True
            else:
                print("Invalid move. Try again.")
        except (ValueError, IndexError):
            print("Invalid input. Enter coordinates as 'start_x start_y end_x end_y'")
    return False

# Main game loop
def play_checkers():
    board = initialize_board()
    print("Welcome to Checkers!")
    print_board(board)
    while True:
        if not user_turn(board):
            print("No valid moves left. You lose!")
            break
        print_board(board)
        if not computer_turn(board):
            print("No valid moves left for the computer. You win!")
            break
        print_board(board)

if __name__ == "__main__":
    play_checkers()

File: src/Checkers.java
import java.util.Scanner;

public class Checkers {

    private static final int BOARD_SIZE = 8;
    private static final char EMPTY = '.';
    private static final char PLAYER_ONE = 'X';
    private static final char PLAYER_TWO = 'O';

    private char[][] board;

    public Checkers() {
        board = new char[BOARD_SIZE][BOARD_SIZE];
        initializeBoard();
    }

    private void initializeBoard() {
        for (int i = 0; i < BOARD_SIZE; i++) {
            for (int j = 0; j < BOARD_SIZE; j++) {
                if ((i + j) % 2 != 0) {
                    if (i < 3) {
                        board[i][j] = PLAYER_ONE;
                    } else if (i > 4) {
                        board[i][j] = PLAYER_TWO;
                    } else {
                        board[i][j] = EMPTY;
                    }
                } else {
                    board[i][j] = EMPTY;
                }
            }
        }
    }

    private void printBoard() {
        System.out.println("  0 1 2 3 4 5 6 7");
        for (int i = 0; i < BOARD_SIZE; i++) {
            System.out.print(i + " ");
            for (int j = 0; j < BOARD_SIZE; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
    }

    private boolean isValidMove(int startX, int startY, int endX, int endY, char player) {
        if (endX < 0 || endX >= BOARD_SIZE || endY < 0 || endY >= BOARD_SIZE) {
            return false;
        }
        if (board[endX][endY] != EMPTY) {
            return false;
        }
        int direction = (player == PLAYER_ONE) ? 1 : -1;
        if (Math.abs(startX - endX) == 1 && endY - startY == direction) {
            return true;
        }
        if (Math.abs(startX - endX) == 2 && Math.abs(endY - startY) == 2) {
            int midX = (startX + endX) / 2;
            int midY = (startY + endY) / 2;
            if (board[midX][midY] != EMPTY && board[midX][midY] != player) {
                return true;
            }
        }
        return false;
    }

    private void makeMove(int startX, int startY, int endX, int endY, char player) {
        board[startX][startY] = EMPTY;
        board[endX][endY] = player;
        if (Math.abs(startX - endX) == 2) {
            int midX = (startX + endX) / 2;
            int midY = (startY + endY) / 2;
            board[midX][midY] = EMPTY;
        }
    }

    public void play() {
        Scanner scanner = new Scanner(System.in);
        char currentPlayer = PLAYER_ONE;
        while (true) {
            printBoard();
            System.out.println("Player " + currentPlayer + "'s turn.");
            System.out.print("Enter startX startY endX endY: ");
            int startX = scanner.nextInt();
            int startY = scanner.nextInt();
            int endX = scanner.nextInt();
            int endY = scanner.nextInt();

            if (isValidMove(startX, startY, endX, endY, currentPlayer)) {
                makeMove(startX, startY, endX, endY, currentPlayer);
                currentPlayer = (currentPlayer == PLAYER_ONE) ? PLAYER_TWO : PLAYER_ONE;
            } else {
                System.out.println("Invalid move. Try again.");
            }
        }
    }

    public static void main(String[] args) {
        Checkers game = new Checkers();
        game.play();
    }
}